### ISSUE TITLE
[feature]: store user sessions in DB

### ISSUE THREAD

**Issue by RaHehl** [hoppscotch/hoppscotch#4379]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

We would like to use pod scaling, as far as I understand it would be necessary for all (api) pods to have the same session information, so the best solution would probably be to store the sessions in the database.

Maybe this can be easily solved: https://github.com/kleydon/prisma-session-store

### Why should this be worked on?

Makes Hoppscotch more enterprise-ready

**Comment by sah246ir** [comment]
i am really interested in working on this issue. can you assign it to me ?

**Comment by yashs33244** [comment]
Hey there, if no one's working on this, can i be assigned this issue? Thank you!

**Comment by iRittikSharma** [comment]
Hi @RaHehl I would like to work on this issue. Could You please assign it to me

**Comment by RaHehl** [comment]
Hi, I'm just the creator of the "issue". From what I understand, this should be assigned by someone from the Hopscotch team? @liyasthomas

**Comment by liyasthomas** [comment]
Hi @iRittikSharma, do you have a proposal for the implementation. I'd love to discuss different possibilities of the implementation with Hoppscotch development team.

Feel free to open up a draft PR if you've already started working on it.

**Comment by sah246ir** [comment]
here's my workflow for the issue

- firstly we can create a database table/collection to store the session id's and a few more parameters if required, this can be done from scratch on login or using the prisma-session-store library.
- secondly we will store the session id as a secure http only cookie on the client side which will be attached to every request from the client.
- thirdly we can create or modify existing  middlewares to validate authentication.
- lastly we can create or modify an endpoint to retrieve authentication status. the client will ping this route once on login to confirm the auth.

p.s @liyasthomas it would be better if you could assign it to someone so that there is only one person responsible for the task.

i would love to work on this too.

**Comment by balub** [comment]
Hey, I can work on this.

**Comment by balub** [comment]
@RaHehl correct me if I am wrong here, but we essentially need to implement basic session storage now, where the session info is stored in our postgres?

**Comment by AndrewBastin** [comment]
@RaHehl The self-hosted backend server uses JWTs to encode session information, so I am kinda confused by the session point you are raising here, the authorisation is done by reading the token, not across the DB or a session state that we maintain.

The challenge to pod scaling (horizontal scaling) implementation is actually related to mostly regarding two things.

1. Data Subscriptions - The API server talks over GraphQL to listen to changes happening to workspaces, this is how realtime updates of collections and environments work on Hoppscotch. When you have multiple backend instances running, updates to the data could be accessing in one server and not the other.
2. Configuration Changes - The admin dashboard allows you to set configuration updates that needs the server to restart. Notifying other instances to do that is a challenge as well.

Support for horizontal scaling is something we are working on and we will aim to ship it by the end of this month for the 2024.10 release cycle.

@RaHehl, can you confirm that you meant this issue to raise support for horizontal scaling (ability to have multiple instances of AIO/Backend containers to handle scaled loads ?) and not something else ?

**Comment by RaHehl** [comment]
@AndrewBastin yes i can confirm it's for horizontal scaling, scaling loads and maybe version upgrades without downtime.

From a security perspective I would like to understand the jwt theme better
